### DISCUSSION TITLE
How do I return a file as response?

### DISCUSSION THREAD

**Discussion by snovey** [vercel/next.js#52588] [App Router] [answered]
Here I have look up some related questions, but still do not solve my doubts:
* [How do I return a file in response using NextResponse? ](https://github.com/vercel/next.js/discussions/50078)
* [Send file as response](https://github.com/vercel/next.js/discussions/15453)

From Next.js 13 it provide the App Router, therefore some old methods cannot apply to this mode.

Since our product is a UGC website, the sitemap is updated daily, and the sitemap is placed on cloud storage, so I want to request a file as a response (there is no direct forwarding because I don’t see which website’s sitemap will use 307 code to forward).

If I write these code at `src/app/sitemap.ts`:
```typescript
export default async function Sitemap(req: Request, res: Response) {
  return await fetch('https://www.oss-cloud.com/sitemap.xml');
}

```
It will report an error because I must return JSX instead of response.

I know I can do this by middleware, but I do not think middleware is used to do this.

```typescript
export async function middleware() {
  return await fetch('https://www.oss-cloud.com/sitemap.xml');
}

export const config = {
  matcher: ['/sitemap.xml'],
}
```

So what should I do ?

I know how to [fetch data](https://nextjs.org/docs/app/building-your-application/data-fetching/fetching) and return [response](https://nextjs.org/docs/app/api-reference/functions/next-response), but I do not know where the code should be written.

**Comment by icyJoseph** [comment, marked as answer]
Alright, as said on my original (now deleted) comment, you could get away with an API route.

First though, the problem is that the sitemap file, has to return either XML statically defined, or a list of objects with url and last modified keys.

- https://nextjs.org/docs/app/api-reference/file-conventions/metadata/sitemap#generate-a-sitemap

So you could, in principle, parse the result from your fetch call, into text, and then feed that to a NPM package, that could do XML text to JSON parsing. 

I personally don't like that one very much, because we'd need to trust an NPM package in various levels, but it is an option. Once you have the JSON representation of the XML, you could return the list of objects with url and last modified keys.

> One could also roll a custom parser, I guess

From what I can tell, sitemap doesn't receive any meaningful arguments for your purpose either. 

One way would be to rewrite, which was my initial suggestion.

1. `rewrite` in next.config.js
```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  async rewrites() {
    return {
      beforeFiles: [{ source: "/sitemap.xml", destination: "/api/sitemap" }],
    };
  },
};

module.exports = nextConfig;

```

2. create the api handler

```ts
// src/app/api/sitemap/route.ts
export const dynamic = "force-dynamic";

export async function GET() {
  const sitemapResponse = await fetch("https://www.sitemaps.org/sitemap.xml");

  const xml = await sitemapResponse.text();

  const response = new Response(xml, {
    status: 200,
    statusText: "ok",
  });

  response.headers.append("content-type", "text/xml");

  return response;
}

```

And tada!  The downside being that there's an endpoint now, `/api/sitemap` that is also reachable. Which is why, middleware could also help here, by rejecting direct access to `api/sitemap`, and only rewriting `/sitemap.xml`.

```ts
// middleware.ts

import { NextRequest, NextResponse } from "next/server";

export const middleware = (req: NextRequest) => {
  if (req.nextUrl.pathname === "/api/sitemap") {
    return new NextResponse(JSON.stringify({ message: "Not Found" }), {
      status: 404,
      headers: { "content-type": "application/json" },
    });
  }

  if (req.nextUrl.pathname === "/sitemap.xml") {
    req.nextUrl.pathname = "/api/sitemap";
    return NextResponse.rewrite(req.nextUrl);
  }

  return NextResponse.next();
};

```

> Don't forget to remove the next.config.js rewrite if you are using the middleware instead

This middleware, without a matcher, runs for every request though.

**Reply by snovey** [reply]
Sorry, I leave out some context information. We use Next.js as BFF layer and not connected to database.

This file is generated every day. we only need to forward it. I don’t know why it needs to be parsed and forwarded after the request. I think it is enough to forward directly like this:
```typescript
// src/app/api/sitemap/route.ts

export async function GET() {
  return await fetch("https://www.sitemaps.org/sitemap.xml");
}
```

Although intercepting the api/sitemap would be perfect, it looks like we actually put a static file. But I think it doesn't matter if you don't block it.

Thank you a lot for your careful answer.
